rust/yarp-sys/build.rs

```diff
@@ -0,0 +1,146 @@
+use std::path::{Path, PathBuf};
+
+fn main() {
+    let ruby_build_path = ruby_build_path();
+    let ruby_include_path = ruby_include_path();
+
+    // Tell cargo/rustc that we want to link against `librubyparser.a`.
+    println!("cargo:rustc-link-lib=static=rubyparser");
+
+    // Add `[root]/build/` to the search paths, so it can find `librubyparser.a`.
+    println!(
+        "cargo:rustc-link-search=native={}",
+        ruby_build_path.to_str().unwrap()
+    );
+
+    // This is where the magic happens.
+    let bindings = generate_bindings(&ruby_include_path);
+
+    // Write the bindings to file.
+    write_bindngs(&bindings);
+}
+
+/// Gets the path to project files (`librubyparser*`) at `[root]/build/`.
+///
+fn ruby_build_path() -> PathBuf {
+    cargo_manifest_path()
+        .join("../../build/")
+        .canonicalize()
+        .unwrap()
+}
+
+/// Gets the path to the header files that `bindgen` needs for doing code generation.
+///
+fn ruby_include_path() -> PathBuf {
+    cargo_manifest_path()
+        .join("../../include/")
+        .canonicalize()
+        .unwrap()
+}
+
+fn cargo_manifest_path() -> PathBuf {
+    PathBuf::from(std::env::var_os("CARGO_MANIFEST_DIR").unwrap())
+}
+
+/// Uses `bindgen` to generate bindings to the C API. Update this to allow new types/functions/etc
+/// to be generated (it's allowlisted to only expose functions that'd make sense for public
+/// consumption).
+///
+/// This method only generates code in memory here--it doesn't write it to file.
+///
+fn generate_bindings(ruby_include_path: &Path) -> bindgen::Bindings {
+    bindgen::Builder::default()
+        .derive_default(true)
+        .generate_block(true)
+        .generate_comments(true)
+        .header(ruby_include_path.join("yarp/defines.h").to_str().unwrap())
+        .header(ruby_include_path.join("yarp.h").to_str().unwrap())
+        .clang_arg(format!("-I{}", ruby_include_path.to_str().unwrap()))
+        .clang_arg("-fparse-all-comments")
+        .impl_debug(true)
+        .layout_tests(true)
+        .merge_extern_blocks(true)
+        .parse_callbacks(Box::new(bindgen::CargoCallbacks))
+        .prepend_enum_name(false)
+        .size_t_is_usize(true)
+        .sort_semantically(true)
+        // Structs
+        .allowlist_type(r#"^yp_\w+_node_t"#)
+        .allowlist_type("yp_buffer_t")
+        .allowlist_type("yp_comment_t")
+        .allowlist_type("yp_diagnostic_t")
+        .allowlist_type("yp_encoding_changed_callback_t")
+        .allowlist_type("yp_encoding_decode_callback_t")
+        .allowlist_type("yp_memsize_t")
```

Comment:
I don't think this needs to be exposed.

Reply:
This comment meant to only quote one line, I think the concurrent git push messed it up.
Basically assume I'm only talking about the last line of every comment unless mentioned otherwise.